sql/src/main/java/io/crate/operation/scalar/IfnullFunction.java

```diff
@@ -0,0 +1,193 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.scalar;
+
+import com.google.common.base.Preconditions;
+import io.crate.metadata.*;
+import io.crate.operation.Input;
+import io.crate.planner.symbol.Function;
+import io.crate.planner.symbol.Literal;
+import io.crate.planner.symbol.Symbol;
+import io.crate.types.CollectionType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import org.apache.lucene.util.BytesRef;
+
+import java.util.List;
+import java.util.Locale;
+
+/**
+ *
+ * IfnullFunction class to implement the ifnull(expr1, expr2) database function
+ * (based on and in the style of the ConcatFunction)
+ *
+ * @author: kanisfluh
+ * @version: 0.1
+ *
+ */
+
+public abstract class IfnullFunction extends Scalar<BytesRef, BytesRef> {
+
+    public static final String NAME = "ifnull";
+    private static final BytesRef EMPTY_STRING = new BytesRef("");
+    private FunctionInfo functionInfo;
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(NAME, new Resolver());
+    }
+
+    protected IfnullFunction(FunctionInfo functionInfo) {
+        this.functionInfo = functionInfo;
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return functionInfo;
+    }
+
+
+    @Override
+    public Symbol normalizeSymbol(Function function) {
+        if (anyNonLiterals(function.arguments())) {
+            return function;
+        }
+        Input[] inputs = new Input[function.arguments().size()];
+        for (int i = 0; i < function.arguments().size(); i++) {
+            inputs[i] = ((Input) function.arguments().get(i));
+        }
+        //noinspection unchecked
+        return Literal.newLiteral(functionInfo.returnType(), evaluate(inputs));
+    }
+
+    /**
+     * specific class handling cases where both input parameters of ifnull(expr1, expr2) are strings
+     */
+
+    private static class StringIfnullFunction extends IfnullFunction {
+
+        protected StringIfnullFunction(FunctionInfo functionInfo) {
+            super(functionInfo);
+        }
+
+        @Override
+        public BytesRef evaluate(Input[] args) {
+            BytesRef firstArg = (BytesRef) args[0].value();
+            BytesRef secondArg = (BytesRef) args[1].value();
+
+            // special case: return "(null)" if both arguments are null ...
+            if ((firstArg == null) && (secondArg == null))
+                return new BytesRef("(null)");  // not the EMPTY_STRING ...
```

Comment:
yes, and should be obsolete in next version since we will return the incoming datatypes (if equal) and not strings only any more, ok!